axlearn/common/input_base.py

```diff
@@ -2,23 +2,121 @@
 
 """Base Input interface."""
 
-from typing import Iterable, Iterator, Optional, Sequence, Union
+import re
+from typing import Iterable, Iterator, NamedTuple, Optional, Protocol, Sequence, Union
 
 import jax
+from absl import logging
+from jax._src.mesh import thread_resources
 from jax.sharding import PartitionSpec
 
-from axlearn.common.config import config_class
+from axlearn.common.config import ConfigOr, config_class, maybe_instantiate
 from axlearn.common.input_dispatch import InputDispatcher
 from axlearn.common.module import Module
 from axlearn.common.utils import (
     Nested,
     Tensor,
     as_numpy_array,
     dispatch_input_batch,
+    tree_paths,
     with_sharding_constraint,
 )
 
 
+class InputPartitionFn(Protocol):
+    """Partitions the input batch."""
+
+    def __call__(self, input_batch: Nested[Tensor]) -> Nested[Tensor]:
+        """Applies sharding constraints to `input_batch` and returns the modified batch.
+
+        Implementations should avoid making in-place updates to `input_batch`.
+        """
+
+
+class PathAndRank(NamedTuple):
+    """A tuple (path, rank) used for matching against inputs in a batch.
+
+    Attributes:
+        path: An optional path or path regex. None means match everything.
+        rank: An optional rank (ndim). None means match everything.
+    """
+
+    path: Optional[Union[str, re.Pattern]]
+    rank: Optional[int]
+
+
+def partition_by_path_rank(
+    path_rank_to_partition: dict[PathAndRank, PartitionSpec],
```

Comment:
Nit: maybe we don't really need a dict:

```suggestion
    path_rank_to_partition: Sequence[IntputFieldPartitionSpec],
```

it simplifies the data structure and `IntputFieldPartitionSpec` is more extensible than `PartitionSpec`.

Reply:
Thanks -- to clarify, did you mean to specify `InputFieldPartitionSpec` as a `NamedTuple` of `(path, rank, PartitionSpec)`, or something else?

(To unblock my next PR, I am happy to address this in a follow-up, if that's OK.)

Reply:
> Thanks -- to clarify, did you mean to specify `InputFieldPartitionSpec` as a `NamedTuple` of `(path, rank, PartitionSpec)`

Yes, a namedtuple, a dataclass, or a chex.struct class.

>, or something else?
> 
> (To unblock my next PR, I am happy to address this in a follow-up, if that's OK.)

Sure.